RESTserver/src/main/java/com/softserve/if072/restservice/service/ForecastService.java

```diff
@@ -0,0 +1,316 @@
+    * the one time span like day : daysDiff[i]=timeSpans[i]/MILLISECONDS_PER_DAY.
+    * 4. Calculate the speeds of the product using for each time span in the daysDiff array.
+    * 5. Calculate the main descriptive statistics values for the productUsedSpeeds array : mean, min, max, count, range,
+    * standard deviation, lower and upper three sigma limits.
+    * 6. If standard deviation is not equals to 0, then the product used speeds set could contain outliers in the data and
+    * we should detect them and deal with them. According to the theory of the normal data distribution, 99.73% of the values
+    * lie within three standard deviations of the mean or within the lower and upper three sigma limits. Therefore, we can
+    * convince that the data values that are beyond three standard deviations from the mean (less than the lower three sigma
+    * limit or more than the upper three sigma limit) are outliers with the probability of 0.9973.
+    * 7. Detect the data that are less than the lower three sigma limit or more than the upper three sigma. If there are the
+    * outliers in the data set we can deal with them in two ways:
+    *  - if the outliers occur at the beginning of the data set or at the end of the data set we can simply remove them (trim
+    * the data set);
+    * - if the outliers occur between two certain values we can simply change them to the average of this two certain values.
+    * 8. Calculate the date of the product ending using the mean from step 4 as the average speed.
+    */
+
+    /**
+     * Construct the ProductStatistics object, that stores statistical information about product using
+     *
+     * @param productId - product unique identifier
+     * @return - ProductStatistics object
+     * @throws NotEnoughDataException
+     */
+    public ProductStatistics getProductStatistics(int productId) throws NotEnoughDataException {
+        LOGGER.trace("getProductStatistics method starts to execute");
+        List<HistoryDTO> historyDTOs = historyDAO.getDTOByProductIdAndAction(productId, Action.USED);
+        if (CollectionUtils.isEmpty(historyDTOs) || historyDTOs.size() < 2) {
+            LOGGER.info("List of product using size is {}. NotEnoughDataException will be thrown.", historyDTOs.size());
+
+            String productName = productService.getProductById(productId).getName();
+
+            throw new NotEnoughDataException(String.format("You don't have enough data records about \"%s\"" +
+                    " for successful forecasting its ending date.", productName), productName);
+        }
+
+        ProductStatistics productStatistics = new ProductStatistics();
+        productStatistics.setProductUsingSpeeds(getProductUsedSpeeds(historyDTOs, productStatistics));
+
+        while (true) {
+            productStatistics.setHasOutliers(false);
+            getDescriptiveStatistics(productStatistics);
+
+            if (productStatistics.getStandardDeviation() != 0) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Equality tests should not be made with floating point values. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonarqube.com/coding_rules#rule_key=squid%3AS1244)